Allow override of dup. field error, and zlib compression level

ihnorton opened the pull request:
- Allow override of duplicate field error, with a warning instead (occasionally occur with buggy writers)

- Allow changing the compression level - can significantly improve write performance at minimal space cost. As an illustrative example:


<details>
<summary>timing test details</summary>

```python
import nrrd
nrrd.reader._NRRD_ALLOW_DUPLICATE_FIELD = True
```


```python
img,hdr = nrrd.read("/path/to/dwi.nhdr")
%timeit -n1 -r1 nrrd.write("test.nrrd", img, hdr)
```

    193 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)



```python
hdr['encoding'] = 'gz'
%timeit -n1 -r1 nrrd.write("test_zl_9.nrrd", img, hdr)
```

    20.2 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)

```python
nrrd.writer._ZLIB_LEVEL = -1
hdr['encoding'] = 'gz'
%timeit -n1 -r1 nrrd.write("test_zl_default.nrrd", img, hdr)
```

    9.94 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)



```python
nrrd.writer._ZLIB_LEVEL = 1
hdr['encoding'] = 'gz'
%timeit -n1 -r1 nrrd.write("test_zl_1.nrrd", img, hdr)
```

    2.68 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)

</details>
<br><br>

Output files below (test.nrrd is original, and zl_default is roughly level 6 per the docs):
```python
!ls -lah *.nrrd

    -rw-r--r--  1 inorton  staff    61M Aug 24 16:38 test.nrrd
    -rw-r--r--  1 inorton  staff    27M Aug 24 16:39 test_zl_1.nrrd
    -rw-r--r--  1 inorton  staff    26M Aug 24 16:38 test_zl_9.nrrd
    -rw-r--r--  1 inorton  staff    26M Aug 24 16:38 test_zl_default.nrrd
```

So, level 9 is ~10x slower than level 1, but saves only 1 MB. This is brain MRI data.

addisonElliott commented on nrrd/tests/test_reading.py:
```diff
@@ -100,6 +100,26 @@ def test_read_raw_header(self):
+
+        dup_message = "Duplicate header field: 'type'"
+        try:
+            header = nrrd.read_header(header_txt_tuple)
```
Can make this one line with the *assertRaisesRegex* function. Check here for an example using https://github.com/mhe/pynrrd/blob/master/nrrd/tests/test_writing.py#L178

ihnorton replied:
done

addisonElliott commented on nrrd/tests/test_reading.py:
```diff
@@ -100,6 +100,26 @@ def test_read_raw_header(self):
+        expected_header = {u'type': 'float', u'dimension': 3}
+        header_txt_tuple = ('NRRD0005', 'type: float', 'dimension: 3', 'type: float')
+
+        dup_message = "Duplicate header field: 'type'"
```
Along with comment below, would remove dup_message and just put as string in *assertRaisesRegex* function

ihnorton commented:
> Also, issue with Travis failing on Python 2.7, need to resolve this as well.

Addressed.

> I would make these parameters as apart of the read function instead.

I made the compression level a parameter of `write` and added the docstring.
But the field error override is kind of odd/special, and violates the nrrd spec, so it probably shouldn't be advertised.

addisonElliott commented on nrrd/writer.py:
```diff
@@ -263,7 +267,7 @@ def _write_data(data, fh, header):
-            compressobj = zlib.compressobj(9, zlib.DEFLATED, zlib.MAX_WBITS | 16)
+            compressobj = zlib.compressobj(compression_level, zlib.DEFLATED, zlib.MAX_WBITS | 16)
         elif header['encoding'] in ['bzip2', 'bz2']:
             compressobj = bz2.BZ2Compressor()
```
Add the compression level to parameter here so that it applies for bzip2 as well

ihnorton replied:
Done

addisonElliott commented on nrrd/writer.py:
```diff
@@ -124,6 +124,10 @@ def write(filename, data, header={}, detached_header=False, custom_field_map=Non
     custom_field_map : :class:`dict` (:class:`str`, :class:`str`), optional
         Dictionary used for parsing custom field types where the key is the custom field name and the value is a
         string identifying datatype for the custom field.
+    compression_level : int
```
Use :class:`int` here just so Sphinx will link to it. Just to stay consistent with the other parameters.

ihnorton replied:
Ok, for consistency. But linking docs for native types is excessive IMHO, and it makes the docstrings ugly outside of sphinx.

addisonElliott commented on nrrd/writer.py:
```diff
@@ -124,6 +124,10 @@ def write(filename, data, header={}, detached_header=False, custom_field_map=Non
         Dictionary used for parsing custom field types where the key is the custom field name and the value is a
         string identifying datatype for the custom field.
+    compression_level : int
+        Int specifying compression level, when applicable.
```
Instead of *when applicable*, maybe be more specific and say *when encoding is set to compression for BZIP or GZIP*

ihnorton replied:
Ok

addisonElliott commented on nrrd/reader.py:
```diff
@@ -15,6 +16,10 @@
 
+# Duplicated fields are prohibited by the spec, but do occur in the wild.
+# Set True to allow duplicate fields, with a warning.
+_NRRD_ALLOW_DUPLICATE_FIELD = False
```
I would make this a parameter to the *read* & *read_header* functions. As I said, I'm not a fan of having to edit a private member in a module, one that especially isn't documented.

I understand your apprehension not to document this because it is not supported by the NRRD specification, but I think whatever parameters we add, they should be documented. 

What you can do though, in the docstring, you can put a warning saying it is not supported. Something like this (I'm not sure if it can go in the argument list, but you can put it in the body of the docstring for sure).

```
    .. warning::
        Allowing duplicate fields with :obj:`allow_duplicate_fields` is not explicitly supported in the NRRD specification. Use this parameter at your own risk.
```

ihnorton replied:
Ok. It is no longer a private field, but I will not be adding it to `read`. This really is a special case, just like the existing chunk size parameter.

addisonElliott commented on nrrd/reader.py:
```diff
@@ -250,7 +255,12 @@ def read_header(file, custom_field_map=None):
+            if not _NRRD_ALLOW_DUPLICATE_FIELD:
+                raise NRRDError(dup_message)
+
+            warnings.warn(dup_message)
```
So with a duplicated field, the old field will be overwritten. Would it be better if we added an integer to the end of the field. Something like <field>2 so that it doesn't conflict but both of the values are saved?

ihnorton replied:
I think it's fine to overwrite. If the values were duplicated then there would probably need to be a hack in the save path to remove the second key while saving, which would just be confusing.

addisonElliott commented:
LGTM 👍 

I guess I wasn't fully thinking of the application for the duplicate field section. I think I agree that it should not be a function parameter in `read` or `read_header`. 

The tests look good!
